File: cpp/173.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <deque>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <numeric>
#include <sstream>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class BSTIterator {
  TreeNode *cur;
  vector<TreeNode*> parents;
public:
  BSTIterator(TreeNode *root) {
    cur = root;
    while(cur && cur->left != nullptr){
      parents.push_back(cur);
      cur = cur->left;
    }
  }

  /** @return whether we have a next smallest number */
  bool hasNext() {
    return cur != nullptr;
  }

  /** @return the next smallest number */
  int next() {
    int v = cur->val;
    if(cur->right != nullptr){
      cur = cur->right;
      while(cur->left != nullptr){
        parents.push_back(cur);
        cur = cur->left;
      }
    } else {
      if(parents.empty())
        cur = nullptr;
      else {
        cur = parents.back();
        parents.pop_back();
      }
    }
    return v;
  }
};

int main(){
  // auto sol = Solution();
  TreeNode a(1), b(2), c(3);
  b.left = &a; b.right = &c;
  BSTIterator i = BSTIterator(nullptr);
  while (i.hasNext()) cout << i.next();
  return 0;
}

File: python/173.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.parents = []
        self.p = root
        while self.p and self.p.left:
            self.parents.append(self.p)
            self.p = self.p.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.p is not None

    def next(self):
        """
        :rtype: int
        """
        val = self.p.val
        if self.p.right:
            self.p = self.p.right
            while self.p.left:
                self.parents.append(self.p)
                self.p = self.p.left
        else:
            if self.parents:
                self.p = self.parents.pop()
            else:
                self.p = None
        return val

    # def next(self):  # not good as above!
    #     """
    #     :rtype: int
    #     """
    #     val = self.p.val
    #     if self.p.right:
    #         self.parents.append(self.p)  # not necessary!
    #         self.p = self.p.right
    #         while self.p.left:
    #             print('left')
    #             self.parents.append(self.p)
    #             self.p = self.p.left
    #     else:
    #         while self.p == self.parents[-1].right:  # not necessary!
    #             self.p = self.parents[-1].pop()
    #         if not self.parents:
    #             self.p = None
    #         else:
    #             self.p = self.parents[-1].pop()
    #     return val

# Your BSTIterator will be called like this:
root = TreeNode(2)
root.left = TreeNode(1)
root.right = TreeNode(3)
i, v = BSTIterator(root), []
while i.hasNext(): v.append(i.next())
